File: python/problems/leetcode/0785_is_graph_bipartite/solution.py
"""LeetCode 785: Is Graph Bipartite?"""

from __future__ import annotations

from collections import deque


class Solution:
    def isBipartite(self, graph: list[list[int]]) -> bool:
        n = len(graph)
        color = [-1] * n

        for start in range(n):
            if color[start] != -1:
                continue
            color[start] = 0
            q: deque[int] = deque([start])
            while q:
                u = q.popleft()
                for v in graph[u]:
                    if color[v] == -1:
                        color[v] = color[u] ^ 1
                        q.append(v)
                    elif color[v] == color[u]:
                        return False
        return True


if __name__ == "__main__":
    assert Solution().isBipartite([[1, 2, 3], [0, 2], [0, 1, 3], [0, 2]]) is False
    assert Solution().isBipartite([[1, 3], [0, 2], [1, 3], [0, 2]]) is True
    assert Solution().isBipartite([[]]) is True
    print("0785 OK")

File: cpp/problems/leetcode/0785_is_graph_bipartite/solution.cpp
// LeetCode 785: Is Graph Bipartite?
#include <alg_std.hpp>
#include <cassert>
using namespace std;

class Solution {
public:
    bool isBipartite(vector<vector<int>>& graph) {
        int n = (int)graph.size();
        vector<int> color(n, -1);
        for (int start = 0; start < n; ++start) {
            if (color[start] != -1) continue;
            color[start] = 0;
            queue<int> q;
            q.push(start);
            while (!q.empty()) {
                int u = q.front();
                q.pop();
                for (int v : graph[u]) {
                    if (color[v] == -1) {
                        color[v] = color[u] ^ 1;
                        q.push(v);
                    } else if (color[v] == color[u]) {
                        return false;
                    }
                }
            }
        }
        return true;
    }
};

int main() {
    vector<vector<int>> g1{{1, 2, 3}, {0, 2}, {0, 1, 3}, {0, 2}};
    assert(!Solution().isBipartite(g1));
    vector<vector<int>> g2{{1, 3}, {0, 2}, {1, 3}, {0, 2}};
    assert(Solution().isBipartite(g2));
    vector<vector<int>> g3(1);
    assert(Solution().isBipartite(g3));
    cout << "0785 OK" << endl;
    return 0;
}
